usr/lib/common/uri.c

```diff
@@ -0,0 +1,268 @@
+        if (((a)[0].type == (t)) &&                      \
+            ((a)[0].pValue)) {                           \
+            p11_buffer_append_len(buf, sep, 1);          \
+            p11_buffer_append(buf, txt);                 \
+            p11_url_encode(START_END((a)[0].pValue,      \
+                                     (a)[0].ulValueLen), \
+                           P11_URI_P_UNRES, buf);        \
+            *sep = P11_URI_SEP_ATTR;                     \
+        }                                                \
+    } while(0)
+
+#define FORMAT_ATTRIBUTE_ALL(sep, txt, a, t, buf)        \
+    do {                                                 \
+        if (((a)[0].type == (t)) &&                      \
+            ((a)[0].pValue)) {                           \
+            p11_buffer_append_len(buf, sep, 1);          \
+            p11_buffer_append(buf, txt);                 \
+            p11_encode(START_END((a)[0].pValue,          \
+                                 (a)[0].ulValueLen),     \
+                       buf);                             \
+            *sep = P11_URI_SEP_ATTR;                     \
+        }                                                \
+    } while(0)
+
+#define FORMAT_ATTRIBUTE_CLASS(sep, txt, a, buf)         \
+    do {                                                 \
+        if (((a)[0].type == CKA_CLASS) &&                \
+            ((a)[0].pValue) &&                           \
+            (c2t((a)[0].pValue))) {                      \
+            p11_buffer_append_len(buf, sep, 1);          \
+            p11_buffer_append(buf, txt);                 \
+            p11_buffer_append(buf, c2t((a)[0].pValue));  \
+            *sep = P11_URI_SEP_ATTR;                     \
+        }                                                \
+    } while(0)
+
+#define FORMAT_VERSION(sep, txt, v, buf)                 \
+    do {                                                 \
+        if (((v).major != (CK_BYTE)-1) ||                \
+            ((v).minor != (CK_BYTE)-1)) {                \
+            p11_buffer_append_len(buf, sep, 1);          \
+            p11_buffer_printf(buf, txt "%d.%d",          \
+                      (v).major, (v).minor);             \
+            *sep = P11_URI_SEP_ATTR;                     \
+        }                                                \
+    } while(0)
+
+#define FORMAT_ULONG(sep, txt, ul, buf)                  \
+    do {                                                 \
+        if ((ul) != (CK_ULONG)-1) {                      \
+            p11_buffer_append_len(buf, sep, 1);          \
+            p11_buffer_printf(buf, txt "%lu", ul);       \
+            *sep = P11_URI_SEP_ATTR;                     \
+        }                                                \
+    } while(0)
+
+static const char HEX_CHARS_LOWER[] = "0123456789abcdef";
+
+static char *c2t(CK_OBJECT_CLASS_PTR class)
```

Comment:
How about moving this function to p11utils.c, i.e. with a different name, e.g. p11_get_cko() ?
This might be useful in other cases, too.

Reply:
Ok, I will move c2t() to p11utils.c. It will also go into a separate commit.

Reply:
I would like to discuss another aspect: the c2t() function is very pkcs11-URI-specific, because the RFC 7512 defines the strings, which are returned by the function. If I would implement it in p11utils.c, I would have selected other strings. e.g.
CKO_CERTIFICATE --> "certificate" (instead of "cert")
CKO_PRIVATE_KEY --> "private-key" (instead of "private")
CKO_PUBLIC_KEY --> "public-key" (instead of "public")

Here is my proposal: I will leave the c2t() function in uri.c but rename it to rfc7512_get_cko(). I will also provide a p11_get_cko() function in p11utils.c.

Reply:
yes that makes a lot of sense. 
If nobody is currently using p11_get_cko(), then you may even decide to not provide it in p11utils.c at the moment, and just have rfc7512_get_cko() in uri.c.